### PR TITLE
Create permission admin endpoints

### PR DESCRIPTION
**Pull request by SilasBerger**
* Currently no explicit checks for existence of user, student group, etc. (handled automatically through foreign key constraints).
* Separate endpoint groups for `/permissions/user` and `permissions/studentGroup`: gets a bit ugly if we mix these things in a single `/permissions` group.
* Endpoints simply return the created, updated or deleted permission model.

### CODE REVIEW

File: src/controllers/rootGroupPermissions.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+export const create: RequestHandler<
+    any,
+    any,
+    { documentRootId: string; studentGroupId: string; access: Access }
+> = async (req, res, next) => {
+    try {
+        const { documentRootId, studentGroupId, access } = req.body;
+
+        if (!(documentRootId && studentGroupId && access)) {
+            res.status(400).send('Missing documentRootId, studentGroupId or access');
```
Comment (lebalz) [inline src/controllers/rootGroupPermissions.ts:14]: Fail loud and let the error handler handle the rest. Import the HTTP400Error from `../utils/errors/Errors`
```suggestion
            throw new HTTP400Error('Missing documentRootId, studentGroupId or access');
```

File: src/controllers/rootGroupPermissions.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        res.status(200).json(model);
+    } catch (error) {
+        next(error);
+    }
+};
+
+export const update: RequestHandler<{ id: string }, any, { access: Access }> = async (req, res, next) => {
+    try {
+        const model = await RootGroupPermission.updateModel(req.params.id, req.body.access);
+        res.status(200).send(model);
```
Comment (lebalz) [inline src/controllers/rootGroupPermissions.ts:29]: notify the users...
```suggestion
        res.status(200).send(model);
        res.notifications = [
            {
                event: IoEvent.CHANGED_RECORD,
                message: { type: RecordType.GroupPermission, record: model },
                to: model.studentGroupId
            }
        ]
```

File: src/controllers/rootGroupPermissions.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        const { documentRootId, studentGroupId, access } = req.body;
+
+        if (!(documentRootId && studentGroupId && access)) {
+            res.status(400).send('Missing documentRootId, studentGroupId or access');
+            return;
+        }
+
+        const model = await RootGroupPermission.createModel(documentRootId, studentGroupId, access);
+
+        res.status(200).json(model);
```
Comment (lebalz) [inline src/controllers/rootGroupPermissions.ts:20]: Notify the users...
```suggestion
        res.status(200).json(model);
        res.notifications = [
            {
                event: IoEvent.NEW_RECORD,
                message: { type: RecordType.GroupPermission, record: model },
                to: model.studentGroupId
            }
        ]
```

File: src/controllers/rootGroupPermissions.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        res.status(200).send(model);
+    } catch (error) {
+        next(error);
+    }
+};
+
+export const destroy: RequestHandler<{ id: string }> = async (req, res, next) => {
+    try {
+        const permission = await RootGroupPermission.deleteModel(req.params.id);
+        res.json(permission);
```
Comment (lebalz) [inline src/controllers/rootGroupPermissions.ts:38]: notify the users
```suggestion
        res.json(permission);
        res.notifications = [
            {
                event: IoEvent.DELETED_RECORD,
                message: { type: RecordType.GroupPermission, record: model },
                to: model.studentGroupId
            }
        ]
```

File: src/controllers/rootUserPermissions.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+export const create: RequestHandler<
+    any,
+    any,
+    { documentRootId: string; userId: string; access: Access }
+> = async (req, res, next) => {
+    try {
+        const { documentRootId, userId, access } = req.body;
+
+        if (!(documentRootId && userId && access)) {
+            res.status(400).send('Missing documentRootId, userId or access');
```
Comment (lebalz) [inline src/controllers/rootUserPermissions.ts:14]: Fail loud..

File: src/controllers/rootUserPermissions.ts [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        res.status(200).json(model);
+    } catch (error) {
+        next(error);
+    }
+};
+
+export const update: RequestHandler<{ id: string }, any, { access: Access }> = async (req, res, next) => {
+    try {
+        const model = await RootUserPermission.updateModel(req.params.id, req.body.access);
+        res.status(200).send(model);
```
Comment (lebalz) [inline src/controllers/rootUserPermissions.ts:29]: notify users...

File: src/controllers/rootUserPermissions.ts [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    } catch (error) {
+        next(error);
+    }
+};
+
+export const destroy: RequestHandler<{ id: string }> = async (req, res, next) => {
+    try {
+        const permission = await RootUserPermission.deleteModel(req.params.id);
```
Comment (lebalz) [inline src/controllers/rootUserPermissions.ts:37]: notify users...

File: src/routes/router.ts [hunk 8]
```diff
@@ -34,6 +44,14 @@ router.get('/studentGroups/:id', findStudentGroup);
 router.put('/studentGroups/:id', updateStudentGroup);
 router.delete('/studentGroups/:id', deleteStudentGroup);
 
+router.post('/permissions/user', createUserPermission);
+router.put('/permissions/user/:id', updateUserPermission);
+router.delete('/permissions/user/:id', deleteUserPermission);
+
+router.post('/permissions/studentGroup', createStudentGroupPermission);
+router.put('/permissions/studentGroup/:id', updateStudentGroupPermission);
+router.delete('/permissions/studentGroup/:id', deleteStudentGroupPermission);
```
Comment (lebalz) [inline src/routes/router.ts:53]: could we simplify this to
```suggestion
router.post('/permissions/group', createStudentGroupPermission);
router.put('/permissions/group/:id', updateStudentGroupPermission);
router.delete('/permissions/group/:id', deleteStudentGroupPermission);
```

And i'm not sure, should it be plural too? `permissions/studentGroups/:id` or `permissions/groups/:id`?

**Review by lebalz — COMMENTED** [review]
Nice job.
I'm wondering wheter we should return the raw db record, or if we should be compliant with https://github.com/GBSL-Informatik/teaching-api/blob/main/src/models/DocumentRoot.ts#L41-L51 and define instead something like:

```ts
interface ApiUserPermission {
        userId: string;
        access: Access;
        documentRootId: string;
}
interface ApiGroupPermission {
        groupId: string;
        access: Access;
        documentRootId: string;
}
```

Then we can use this interface too in DocumentRoot:

```ts

export interface Config {
    access?: Access; // Access level of document root
    userPermissions?: Omit<ApiUserPermission, 'documentRootId'>[];
    groupPermissions?: Omit<ApiGroupPermission, 'documentRootId'>[];
}
```

**Comment by SilasBerger** [comment]
> Nice job. I'm wondering wheter we should return the raw db record, or if we should be compliant with https://github.com/GBSL-Informatik/teaching-api/blob/main/src/models/DocumentRoot.ts#L41-L51 and define instead something like:
> 
> ```ts
> interface ApiUserPermission {
>         userId: string;
>         access: Access;
>         documentRootId: string;
> }
> interface ApiGroupPermission {
>         groupId: string;
>         access: Access;
>         documentRootId: string;
> }
> ```
> 
> Then we can use this interface too in DocumentRoot:
> 
> ```ts
> export interface Config {
>     access?: Access; // Access level of document root
>     userPermissions?: Omit<ApiUserPermission, 'documentRootId'>[];
>     groupPermissions?: Omit<ApiGroupPermission, 'documentRootId'>[];
> }
> ```

Good idea! Turns out that we already had the `ApiUserPermission` and `ApiGroupPermission`, so I am using them in the controllers and the `Config` now.

File: src/models/DocumentRoot.ts [hunk 9]
```diff
@@ -9,7 +9,6 @@ import {
     User
 } from '@prisma/client';
 import { ApiDocument, prepareDocument } from './Document';
-import { highestAccess } from '../helpers/accessPolicy';
 
 export type ApiGroupPermission = {
```
Comment (lebalz) [inline src/models/DocumentRoot.ts:13]: i think `ApiGroupPermission` and `ApiUserPermission` could be moved over to the according model files?

File: src/models/RootGroupPermission.ts [hunk 10]
```diff
@@ -0,0 +1,39 @@
+import { Access, PrismaClient } from '@prisma/client';
+import prisma from '../prisma';
+
+// TODO: Consider checking existence of documentRoot / studentGroup to provide better error messages / exceptions.
+
+function RootGroupPermission(db: PrismaClient['rootGroupPermission']) {
```
Comment (lebalz) [inline src/models/RootGroupPermission.ts:6]: should we return in each method a `ApiGroupPermission`?

File: src/models/RootUserPermission.ts [hunk 11]
```diff
@@ -0,0 +1,39 @@
+import { Access, PrismaClient } from '@prisma/client';
+import prisma from '../prisma';
+
+// TODO: Consider checking existence of documentRoot / user to provide better error messages / exceptions.
+
+function RootUserPermission(db: PrismaClient['rootUserPermission']) {
```
Comment (lebalz) [inline src/models/RootUserPermission.ts:6]: should we return in each method a `ApiUserPermission`?

File: src/models/RootUserPermission.ts [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+
+// TODO: Consider checking existence of documentRoot / user to provide better error messages / exceptions.
+
+export type ApiUserPermission = {
+    id: string;
+    userId: string;
+    access: Access;
+};
+
+function asApiRecord(dbResult: DbRootUserPermission): ApiUserPermission {
```
Comment (lebalz) [inline src/models/RootUserPermission.ts:13]: Good function name, i like it more than the `prepare<ModelName>`.  We should change on a follow up pr all to `asApiRecord`👍

**Review by lebalz — APPROVED** [review]
LGTM, thanks 👌

**Pull request merged**